ouroboros-consensus-test/test-consensus/Test/Consensus/Mempool/StateMachine.hs

```diff
@@ -0,0 +1,810 @@
+       , initialState
+       )
+  where
+    go (acc, tk, cap, st) [] = ((reverse acc, [])
+                               , tk
+                               , MempoolCapacityBytes cap
+                               , st
+                               )
+    go (acc, tk, cap, st) txs@(tx:next) =
+      if cap < txInBlockSize tx
+      then ((reverse acc, txs), tk, MempoolCapacityBytes cap, st)
+      else
+        let slot = case getTipSlot st of
+              Origin -> minimumPossibleSlotNo (Proxy @blk)
+              At v   -> v + 1
+        in
+        case runExcept $ applyTx cfg DoNotIntervene slot tx st of
+          Left e           ->
+            go ( MempoolTxRejected tx e:acc
+               , tk
+               , cap
+               , st
+               )
+               next
+          Right (st', vtx) ->
+            go ( MempoolTxAdded vtx:acc
+               , succ tk
+               , cap - txInBlockSize tx
+               , forgetLedgerTablesDiffsTicked st'
+               )
+               next
+
+tick ::
+  ( TickedTableStuff (LedgerState blk)
+  , ValidateEnvelope blk
+  , LedgerSupportsMempool blk
+  )
+  => LedgerConfig blk
+  -> LedgerState blk ValuesMK
+  -> TickedLedgerState blk ValuesMK
+tick cfg st =
+  zipOverLedgerTablesTicked
+    (flip rawApplyDiffs)
+    ticked
+    (projectLedgerTables st)
+  where
+    ticked = snd
+           . tickLedgerState cfg
+           . ForgeInUnknownSlot
+           . forgetLedgerTables
+           $ st
+
+{-------------------------------------------------------------------------------
+  SUT side
+-------------------------------------------------------------------------------}
+
+-- | The System Under Test
+data SUT m blk =
+  SUT
+  !(Mempool m blk)
+    -- ^ A Mempool
+  !(StrictTVar m (Tips blk))
+    -- ^ A TVar with tips. Emulates a ledger db to the extent needed by the
+    -- ledger interface.
+  deriving Generic
+
+deriving instance ( NoThunks (Mempool m blk)
+                  , NoThunks (StrictTVar m (Tips blk))
+                  ) =>  NoThunks (SUT m blk)
+
+-- | The tip of the ledger db and the one from the mempool. The ledger interface
+-- will serve the values of the latter if the mempool requests it and is
+-- reachable.
+data Tips blk = Tips {
+    ldbTip     :: !(LedgerState blk ValuesMK)
+  , mempoolTip :: !(LedgerState blk ValuesMK, Bool)
+  } deriving (Generic)
+
+-- | Create a ledger interface and provide the tvar to modify it when switching
+-- ledgers.
+newLedgerInterface ::
+  ( MonadSTM m
+  , NoThunks (Tips blk)
+  , GetTip (LedgerState blk ValuesMK)
+  , LedgerSupportsMempool blk
+  )
+  => LedgerState blk ValuesMK
+  -> m (LedgerInterface m blk, StrictTVar m (Tips blk))
+newLedgerInterface initialLedger = do
+  t <- newTVarIO $ Tips initialLedger (initialLedger, True)
+  pure $ (,t) $ LedgerInterface {
+      getCurrentLedgerState = forgetLedgerTables . ldbTip <$> readTVar t
+    , getLedgerTablesAtFor  = \pt txs -> do
```

Comment:
Maybe we should put this outside, in a function and add some comments on what is going on, since it might be a bit hard to follow.

Reply:
I added some comments. I don't think this is difficult to follow at least now that the comments are in place